localstack/http/router.py

```diff
@@ -462,23 +373,223 @@ def route(
+        :param kwargs: any other argument that can be passed to ``werkzeug.routing.Rule``
+        """
+        ...
+
+    @overload
+    def __init__(self, fn: _RouteEndpoint):
+        """
+        Takes a route endpoint (typically a function decorated with ``@route``) and adds it as ``EndpointRule``.
+
+        :param fn: the RouteEndpoint function
+        """
+        ...
+
+    @overload
+    def __init__(self, rule_factory: RuleFactory):
+        """
+        Adds a ``Rule`` or the rules created by a ``RuleFactory`` to the given router. It passes the rules down to
+        the underlying Werkzeug ``Map``, but also returns the created Rules.
+
+        :param rule_factory: a `Rule` or ``RuleFactory`
         """
-        return self._add_route(fn)
+        ...
+
+    @overload
+    def __init__(self, obj: Any):
+        """
+        Scans the given object for members that can be used as a `RouteEndpoint` and adds them to the router.
 
-    def add_route_endpoints(self, obj: object) -> List[Rule]:
+        :param obj: the object to scan
         """
-        DEPRECATED: use ``add`` instead.
+        ...
+
+    @overload
+    def __init__(self, rules: list[Union[_RouteEndpoint, RuleFactory, Any]]):
+        """Add multiple rules at once"""
+        ...
+
+    def __init__(self, *args, **kwargs):
         """
-        return self._add_routes(obj)
+        Dispatcher for overloaded ``__init__`` methods.
+        """
+        if "path" in kwargs or type(args[0]) is str:
+            self.factory = _EndpointRule(*args, **kwargs)
+        elif "fn" in kwargs or callable(args[0]):
+            self.factory = _EndpointFunction(*args, **kwargs)
+        elif "rule_factory" in kwargs:
+            self.factory = kwargs["rule_factory"]
+        elif isinstance(args[0], RuleFactory):
+            self.factory = args[0]
+        elif isinstance(args[0], list):
+            self.factory = RuleGroup([RuleAdapter(rule) for rule in args[0]])
+        else:
+            self.factory = _EndpointsObject(*args, **kwargs)
+
+    def get_rules(self, map: Map) -> t.Iterable[Rule]:
+        yield from self.factory.get_rules(map)
+
+
+class WithHost(RuleFactory):
+    def __init__(self, host: str, rules: Iterable[RuleFactory]) -> None:
+        self.host = host
+        self.rules = rules
+
+    def get_rules(self, map: Map) -> t.Iterator[Rule]:
+        for rulefactory in self.rules:
+            for rule in rulefactory.get_rules(map):
+                rule = rule.empty()
+                rule.host = self.host
+                yield rule
+
+
+class RuleGroup(RuleFactory):
+    def __init__(self, rules: Iterable[RuleFactory]):
+        self.rules = rules
+
+    def get_rules(self, map: Map) -> t.Iterable[Rule]:
+        for rule in self.rules:
+            yield from rule.get_rules(map)
+
 
-    def remove_rule(self, rule: Rule):
-        """DEPRECATED: use ``remove`` instead."""
-        self._remove_rules([rule])
+class _EndpointRule(RuleFactory, Generic[E]):
```

Comment:
This is a nice encapsulation of the creation of rules based on the rule attributes (which was previously in the `_add_endpoint` function). Maybe you could also add a small docstring here?